code/game/objects/structures/beds_chairs/bed.dm

```diff
@@ -18,65 +19,155 @@
 	resistance_flags = FLAMMABLE
 	max_integrity = 100
 	integrity_failure = 0.35
-	var/buildstacktype = /obj/item/stack/sheet/iron
-	var/buildstackamount = 2
-	var/bolts = TRUE
+	/// What material this bed is made of
+	var/build_stack_type = /obj/item/stack/sheet/iron
+	/// How many mats to drop when deconstructed
+	var/build_stack_amount = 2
+	/// If someone's buckled to it
+	var/occupied = FALSE
 
 /obj/structure/bed/Initialize(mapload)
 	. = ..()
 	AddElement(/datum/element/soft_landing)
+	register_context()
 
 /obj/structure/bed/examine(mob/user)
 	. = ..()
-	if(bolts)
+	if(!(flags_1 & NODECONSTRUCT_1))
 		. += span_notice("It's held together by a couple of <b>bolts</b>.")
 
+/obj/structure/bed/add_context(atom/source, list/context, obj/item/held_item, mob/living/user)
+	if(held_item)
+		if(held_item.tool_behaviour != TOOL_WRENCH || flags_1 & NODECONSTRUCT_1)
+			return
+
+		context[SCREENTIP_CONTEXT_RMB] = "Dismantle"
+		return CONTEXTUAL_SCREENTIP_SET
+
+	else if(occupied)
+		context[SCREENTIP_CONTEXT_LMB] = "Unbuckle"
+		return CONTEXTUAL_SCREENTIP_SET
+
 /obj/structure/bed/deconstruct(disassembled = TRUE)
 	if(!(flags_1 & NODECONSTRUCT_1))
-		if(buildstacktype)
-			new buildstacktype(loc,buildstackamount)
+		if(build_stack_type)
+			new build_stack_type(loc, build_stack_amount)
 	..()
 
 /obj/structure/bed/attack_paw(mob/user, list/modifiers)
 	return attack_hand(user, modifiers)
 
 /obj/structure/bed/wrench_act_secondary(mob/living/user, obj/item/weapon)
-	if(flags_1&NODECONSTRUCT_1)
+	if(flags_1 & NODECONSTRUCT_1)
 		return TRUE
+
 	..()
 	weapon.play_tool_sound(src)
 	deconstruct(disassembled = TRUE)
 	return TRUE
 
-/*
- * Roller beds
- */
-/obj/structure/bed/roller
-	name = "roller bed"
-	icon = 'icons/obj/medical/rollerbed.dmi'
-	icon_state = "down"
+/obj/structure/bed/post_buckle_mob(mob/living/buckled_mob)
+	occupied = TRUE
+
+/obj/structure/bed/post_unbuckle_mob(mob/living/unbuckled_mob)
+	occupied = FALSE
+
+/// Medical beds
+
+/obj/structure/bed/medical
+	name = "medical bed"
+	icon = 'icons/obj/medical/medical_bed.dmi'
+	desc = "A medical bed with wheels for assisted patient movement or medbay racing tournaments."
+	icon_state = "med_down"
+	base_icon_state = "med"
 	anchored = FALSE
 	resistance_flags = NONE
-	///The item it spawns when it's folded up.
-	var/foldabletype = /obj/item/roller
+	build_stack_type = /obj/item/stack/sheet/mineral/titanium
+	build_stack_amount = 1
```

Comment:
People who know mats, how likely is this to become an infinite material exploit? These can be printed as well for 1,5 materials, but I'm not entirely sure how the conversions works at tier 4

Reply:
Thanks, I forgot to check that. I raised the material costs so that even at tier 4 you can't exploit it.